src/Optimizer/HeroCandidateFiltering.php

```diff
@@ -0,0 +1,131 @@
+<?php
+/**
+ * Class HeroCandidateFiltering.
+ *
+ * @package AmpProject\AmpWP
+ */
+
+namespace AmpProject\AmpWP\Optimizer;
+
+use AmpProject\AmpWP\Infrastructure\Conditional;
+use AmpProject\AmpWP\Infrastructure\Delayed;
+use AmpProject\AmpWP\Infrastructure\Registerable;
+use AmpProject\AmpWP\Infrastructure\Service;
+use AmpProject\Attribute;
+use WP_Post;
+
+/**
+ * Service which uses WordPress hooks to inject data-hero-candidate attributes on logos, headers, and the primary featured image.
+ *
+ * @package AmpProject\AmpWP
+ * @since 2.1.0
+ * @internal
+ */
+final class HeroCandidateFiltering implements Service, Delayed, Conditional, Registerable {
+
+	/**
+	 * Whether the Custom Logo has been filtered.
+	 *
+	 * This is set to true after the first filtering to prevent filtering the logo in the footer, for example.
+	 *
+	 * @var bool
+	 */
+	protected $custom_logo_filtered = false;
+
+	/**
+	 * Get the action to use for registering the service.
+	 *
+	 * @return string Registration action to use.
+	 */
+	public static function get_registration_action() {
+		return 'wp';
+	}
+
+	/**
+	 * Check whether the conditional object is currently needed.
+	 *
+	 * @return bool Whether the conditional object is needed.
+	 */
+	public static function is_needed() {
+		return amp_is_request();
+	}
+
+	/**
+	 * Register.
+	 */
+	public function register() {
+		add_filter( 'get_custom_logo_image_attributes', [ $this, 'add_custom_logo_data_hero_candidate_attribute' ] );
+		add_filter( 'get_header_image_tag', [ $this, 'filter_header_image_tag' ] );
+		add_filter( 'wp_get_attachment_image_attributes', [ $this, 'filter_attachment_image_attributes' ], 10, 2 );
```

Comment:
For each of these filters we could check if the `data-hero-candidate` attribute exists before modifying anything.

Reply:
I don't think that is necessary. For the `get_custom_logo_image_attributes` and `wp_get_attachment_image_attributes` filters, it would just replace the existing array item with the same value. For `get_header_image_tag` it would just be adding a duplicate attribute which would just be ignored when parsing.

Reply:
Yes that makes sense, thanks for the explanation.